internal/statestorage/redis/redispb/player.go

```diff
@@ -0,0 +1,177 @@
+	player.Pool = playerMap["pool"]
+	player.Assignment = playerMap["assignment"]
+	player.Status = playerMap["status"]
+	player.Error = playerMap["error"]
+
+	// TODO: Room for improvement here.
+	if a := playerMap["attributes"]; a != "" {
+		attrsJSON := fmt.Sprintf("{\"attributes\": %v}", a)
+		err = jsonpb.UnmarshalString(attrsJSON, player)
+		if err != nil {
+			resultLog.Error("failure on attributes")
+			resultLog.Error(a)
+		}
+	}
+
+	resultLog.Debug("state storage operation: player unmarshalled")
+	return err
+
+}
+
+// PlayerWatcher makes a channel and returns it immediately.  It also launches an
+// asynchronous goroutine that watches a redis key and returns updates to
+// that key on the channel.
+//
+// The pattern for this function is from 'Go Concurrency Patterns', it is a function
+// that wraps a closure goroutine, and returns a channel.
+// reference: https://talks.golang.org/2012/concurrency.slide#25
+//
+// NOTE: this function will never stop querying Redis during normal operation! You need to
+//  disconnect the client from the frontend API (which closes the context) once
+//  you've received the results you were waiting for to stop doing work!
+func PlayerWatcher(ctx context.Context, pool *redis.Pool, pb om_messages.Player) <-chan om_messages.Player {
+
+	pwLog := pLog.WithFields(log.Fields{"playerId": pb.Id})
+
+	// Establish channel to return results on.
+	watchChan := make(chan om_messages.Player)
+
+	go func() {
+		// var declaration
+		var prevResults = ""
+
+		// Loop, querying redis until this key has a value or the Redis query fails.
+		for {
+			select {
+			case <-ctx.Done():
+				// Player stopped asking for updates; clean up
+				close(watchChan)
+				return
+			default:
+				// Update the player's 'accessed' timestamp to denote they haven't disappeared
+				err := playerindices.Touch(ctx, pool, pb.Id)
+				if err != nil {
+					// Not fatal, but this error should be addressed.  This could
+					// cause the player to expire while still actively connected!
+					pwLog.Error("State storage error:", err.Error())
+				}
+
+				// Get player from redis.
+				results := om_messages.Player{Id: pb.Id}
+				err = UnmarshalPlayerFromRedis(ctx, pool, &results)
+				if err != nil {
+					// Return error and quit.
+					pwLog.Debug("State storage error:", err.Error())
+					results.Error = err.Error()
+					watchChan <- results
+					close(watchChan)
+					return
+				}
+
+				// Check for new results and send them. Store a copy of the
+				// latest version in string form so we can compare it easily in
+				// future loops.
+				//
+				// If we decide to watch other message fields for updates,
+				// they will need to be added here.
+				//
+				// This can be made much cleaner if protobuffer reflection improves.
+				curResults := fmt.Sprintf("%v%v%v", results.Assignment, results.Status, results.Error)
+				if prevResults == curResults {
+					pwLog.Debug("No new results, backing off")
```

Comment:
Nit. "backing off" implies some kind of change in the wait period. I would be confused if I saw this in the logs. Maybe. "No new watcher results".